mods/carts/cart_entity.lua

```diff
@@ -46,15 +46,19 @@ function cart_entity:on_activate(staticdata, dtime_s)
 		return
 	end
 	self.railtype = data.railtype
-	if data.old_dir then
-		self.old_dir = data.old_dir
+	self.old_dir = data.old_dir or self.old_dir
+	self.old_pos = data.old_pos or self.old_pos
+	-- Correct the position when the cart drives further after the last 'step()'
+	if self.old_pos and carts:is_rail(self.old_pos, self.railtype) then
+		self.object:set_pos(self.old_pos)
 	end
 end
 
 function cart_entity:get_staticdata()
 	return minetest.serialize({
 		railtype = self.railtype,
-		old_dir = self.old_dir
+		old_dir = self.old_dir,
+		old_pos = self.old_pos
```

Comment:
I don't feel quite comfortable with serializing old cart positions here... If data is lost (due to a hard crash), the staticdata will be outdated, potentially causing the cart to teleport. Serializing the direction isn't an issue, because that can be properly updated based on the rails.

Reply:
Good point. How could rail-jumping be avoided then, if not by teleporting to the last safe position? The rail type underneath the cart may chance (or there's no rail at all), resulting in out-of-place carts.

Reply:
> Good point. How could rail-jumping be avoided then, if not by teleporting to the last safe position? The rail type underneath the cart may chance (or there's no rail at all), resulting in out-of-place carts.
> 
> EDIT: I suppose the engine could be blamed in this case.

Well, while the entity step is running, the cart may teleport to the latest rail pos in order to not overshoot. This may only go for a single entity step though and *may not be persisted* in particular to avoid larger teleportation.

Reply:
Whatever. I removed the code.